internal/venice-common/src/main/java/com/linkedin/venice/writer/ChunkHelper.java

```diff
@@ -0,0 +1,162 @@
+      }
+
+      Put putPayload = new Put();
+      putPayload.putValue = chunk;
+      putPayload.schemaId = AvroProtocolDefinition.CHUNK.getCurrentProtocolVersion();
+      putPayload.replicationMetadataVersionId = VENICE_DEFAULT_TIMESTAMP_METADATA_VERSION_ID;
+      putPayload.replicationMetadataPayload = EMPTY_BYTE_BUFFER;
+
+      chunkedKeySuffix.chunkId.chunkIndex = chunkIndex;
+      keyProvider = chunkIndex == 0 ? firstKeyProvider : subsequentKeyProvider;
+
+      try {
+        /**
+         * Here are the reasons to do a blocking call of 'sendMessage' with 'null' callback:
+         * 1. We don't want the upper layer know the chunking logic here. Right now, if we pass the callback parameter,
+         * it will cause the job failure because the message count of sent/completed in 'VeniceReducer' won't match;
+         * 2. Blocking call could guarantee correctness;
+         * 3. Infinite blocking means the following 'sendMessage' call will follow the config of the internal Kafka Producer,
+         * such as timeout, retries and so on;
+         */
+        sendMessageFunction.apply(keyProvider, putPayload).get();
+      } catch (Exception e) {
+        throw new VeniceException(
+            "Caught an exception while attempting to produce a chunk of a large value into Kafka... "
+                + getDetailedSizeReport(chunkIndex, numberOfChunks, sizeAvailablePerMessage, sizeReport),
+            e);
+      }
+    }
+    return new ChunkedPayloadAndManifest(chunks, chunkedValueManifest);
+  }
+
+  private static void validateAvailableSizePerMessage(
+      int maxSizeForUserPayloadPerMessageInBytes,
+      int sizeAvailablePerMessage,
+      Supplier<String> sizeReport) {
+    if (sizeAvailablePerMessage < maxSizeForUserPayloadPerMessageInBytes / 2) {
+      /**
+       * If the key size is more than half the available payload per message, then we cannot even encode a
+       * {@link ChunkedValueManifest} with two keys in it, hence we would fail anyway. Might as well fail fast.
+       *
+       * N.B.: The above may not be strictly true, since Kafka-level compression may squeeze us under the limit,
+       * but this is not likely to work deterministically, so it would be iffy to attempt it on a best-effort
+       * basis. That being said, allowing keys to be as large as half the available payload size is probably
+       * overly permissive anyway. Ideally, the key size should be much smaller than this.
+       *
+       * TODO: Implement a proper key size (and value size) quota mechanism.
+       */
+      throw new VeniceException("Chunking cannot support this use case. The key is too large. " + sizeReport.get());
+    }
+  }
+
+  private static String getDetailedSizeReport(
+      int chunkIndex,
+      int numberOfChunks,
+      int sizeAvailablePerMessage,
+      Supplier<String> sizeReport) {
+    return "Current chunk index: " + chunkIndex + ", " + "Number of chunks: " + numberOfChunks + ", "
```

Comment:
nit: I know it's not your change code, but it'd be good to use `StringBuilder` or `String.format`?